curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -0,0 +1,1092 @@
+            tmp.set_offset(newChunk.offset() + newChunk.len());
+            tmp.set_len(old.offset() + old.len() - newChunk.offset() -
+                        newChunk.len());
+            tmp.set_size(old.offset() + old.len() - newChunk.offset() -
+                         newChunk.len());
+            tmp.set_zero(old.zero());
+            result.push_back(tmp);
+        }
+        /*
+                      -----     old
+                   ----------   new
+        */
+    } else if (newChunk.offset() <= old.offset() &&
+               newChunk.offset() + newChunk.len() >= old.offset() + old.len()) {
+        return result;
+        /*
+                      --------  old
+                   -------      new
+        */
+    } else {
+        tmp.set_chunkid(old.chunkid());
+        tmp.set_compaction(old.compaction());
+        tmp.set_offset(newChunk.offset() + newChunk.len());
+        tmp.set_len(old.offset() + old.len() - newChunk.offset() -
+                    newChunk.len());
+        tmp.set_size(old.offset() + old.len() - newChunk.offset() -
+                     newChunk.len());
+        tmp.set_zero(old.zero());
+        result.push_back(tmp);
+    }
+
+    return result;
+}
+
+void FileCacheManager::ReadChunk(uint64_t index, uint64_t chunkPos,
+                                 uint64_t readLen, char* dataBuf,
+                                 uint64_t dataBufOffset,
+                                 std::vector<ReadRequest>* requests) {
+    ChunkCacheManagerPtr chunkCacheManager = FindChunkCacheManager(index);
+
+    if (!chunkCacheManager) {
+        chunkCacheManager = CreateChunkCacheManager(index);
+    }
+
+    chunkCacheManager->Read(chunkPos, readLen, dataBuf, dataBufOffset,
+                            requests);
+
+    return;
+}
+
+void FileCacheManager::ReleaseCache() {
+    auto iter = chunkCacheMap_.begin();
+    WriteLockGuard writeLockGuard(rwLock_);
+
+    for (; iter != chunkCacheMap_.end(); iter++) {
+        iter->second->ReleaseCache(s3ClientAdaptor_);
+    }
+
+    chunkCacheMap_.clear();
+    return;
+}
+
+CURVEFS_ERROR FileCacheManager::Flush(Inode* inode, bool force) {
+    CURVEFS_ERROR ret;
+    std::map<uint64_t, ChunkCacheManagerPtr> tmp;
+    {
+        WriteLockGuard writeLockGuard(rwLock_);
+        tmp = chunkCacheMap_;
+    }
+    auto iter = tmp.begin();
+
+    for (; iter != tmp.end(); iter++) {
+        ret = iter->second->Flush(inode, force);
+        if (ret != CURVEFS_ERROR::OK) {
+            LOG(ERROR) << "fileCacheManager Flush error, ret:" << ret
+                       << ",chunkIndex:" << iter->second->GetIndex();
+            return ret;
+        }
+        {
+            WriteLockGuard writeLockGuard(rwLock_);
+            auto iter1 = chunkCacheMap_.find(iter->first);
+            if (iter1->second->IsEmpty()) {
+                chunkCacheMap_.erase(iter1);
+            }
+        }
+    }
+    return CURVEFS_ERROR::OK;
+}
+
+CURVEFS_ERROR FileCacheManager::Flush(bool force) {
```

Comment:
This function is not needed， and InodeCacheManager is only used here， it is not needed too

Reply:
but must get inode before call Flush。I don't see the difference